src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesFeature.cs

```diff
@@ -30,6 +33,7 @@ public IServiceProvider RequestServices
                     _scope = _scopeFactory.CreateScope();
                     _requestServices = _scope.ServiceProvider;
                     _requestServicesSet = true;
+                    _context.Response.RegisterForDispose(this);
```

Comment:
This does create one new possibility for a leak. Consider if someone called RequestServices for the first time inside of their RegisterForDispose/OnCompleted callback. It would be too late to register an additional disposal. 

Not a huge risk, the contract of OnCompleted tells you not to touch HttpContext anyways.

Reply:
Yep agreed. Not sure that matters.

Reply:
I could move RegisterForDispose first, that way if it blows up the scope would never be created in the first place.

Reply:
Is this what caused `RequestServices` to be disposed and reset to `null` before the `OnStarting` callbacks are invoked?

Reply:
Umm. Never mind. I see this was done for 2.1. After updating from 2.0.8 to 2.1-rc1-final, I no longer see `RequestServices` being `null` inside `OnStarting`. Weird 🤔